clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3449,39 +3485,49 @@ static bool InvalidWorkGroupSizeAttrs(const Expr *MGValue, const Expr *XDim,
 // __attribute__((reqd_work_group_size)) follows the OpenCL rules in OpenCL
 // mode. All spellings of reqd_work_group_size attribute (regardless of
 // syntax used) follow the SYCL rules when in SYCL mode.
-static bool checkMaxAllowedWorkGroupSize(
-    Sema &S, const Expr *RWGSXDim, const Expr *RWGSYDim, const Expr *RWGSZDim,
+bool Sema::CheckMaxAllowedWorkGroupSize(
+    const Expr *RWGSXDim, const Expr *RWGSYDim, const Expr *RWGSZDim,
     const Expr *MWGSXDim, const Expr *MWGSYDim, const Expr *MWGSZDim) {
   // If any of the operand is still value dependent, we can't test anything.
   const auto *RWGSXDimExpr = dyn_cast<ConstantExpr>(RWGSXDim);
-  const auto *RWGSYDimExpr = dyn_cast<ConstantExpr>(RWGSYDim);
-  const auto *RWGSZDimExpr = dyn_cast<ConstantExpr>(RWGSZDim);
+  const auto *RWGSYDimExpr = dyn_cast_or_null<ConstantExpr>(RWGSYDim);
+  const auto *RWGSZDimExpr = dyn_cast_or_null<ConstantExpr>(RWGSZDim);
   const auto *MWGSXDimExpr = dyn_cast<ConstantExpr>(MWGSXDim);
   const auto *MWGSYDimExpr = dyn_cast<ConstantExpr>(MWGSYDim);
   const auto *MWGSZDimExpr = dyn_cast<ConstantExpr>(MWGSZDim);
 
-  if (!RWGSXDimExpr || !RWGSYDimExpr || !RWGSZDimExpr || !MWGSXDimExpr ||
-      !MWGSYDimExpr || !MWGSZDimExpr)
+  if (!RWGSXDimExpr || (!RWGSYDimExpr && RWGSYDim) ||
+      (!RWGSZDimExpr && RWGSZDim) || !MWGSXDimExpr || !MWGSYDimExpr ||
+      !MWGSZDimExpr)
     return false;
 
-  // Otherwise, check if values of 'reqd_work_group_size' attribute arguments
-  // are greater than values of 'max_work_group_size' attribute arguments.
+  // Y and Z are only optional for SYCL targets.
+  if (!getLangOpts().SYCLIsDevice && !(RWGSYDim && RWGSZDim))
+    return false;
+
+  // SYCL reorders arguments based on the dimensionality.
+  // If we only have the X-dimension, there is no change to the expressions,
+  // otherwise the last specified dimension acts as the first dimension in the
+  // work-group size.
+  const ConstantExpr *FirstRWGDimExpr = RWGSXDimExpr;
+  const ConstantExpr *SecondRWGDimExpr = RWGSYDimExpr;
+  const ConstantExpr *ThirdRWGDimExpr = RWGSZDimExpr;
+  if (getLangOpts().SYCLIsDevice && RWGSYDim)
+    std::swap(FirstRWGDimExpr, RWGSZDim ? ThirdRWGDimExpr : SecondRWGDimExpr);
+
+  // Check if values of 'reqd_work_group_size' attribute arguments are greater
+  // than values of 'max_work_group_size' attribute arguments.
   bool CheckFirstArgument =
-      S.getLangOpts().OpenCL
-          ? RWGSXDimExpr->getResultAsAPSInt().getZExtValue() >
-                MWGSZDimExpr->getResultAsAPSInt().getZExtValue()
-          : RWGSXDimExpr->getResultAsAPSInt().getZExtValue() >
-                MWGSXDimExpr->getResultAsAPSInt().getZExtValue();
+      FirstRWGDimExpr->getResultAsAPSInt().getZExtValue() >
```

Comment:
It generally becomes a little confusing because `max_work_group_size` requires all 3 arguments at the moment, while `reqd_work_group_size` does not. The way SYCL works we swap the first and last dimensions, which means when padding is added to bring the dimensions up to 3, the work-group size given to the backends are actually dependent on the dimensionality of the specified work groups.

So in this code there are potentially two flips; one for `reqd_work_group_size` and one for `max_work_group_size`. Since we know that the latter is always 3-dimensional we compare "first-dim" with the Z of `max_work_group_size` as it is actually the first dimension from a backend-perspective. However, for `reqd_work_group_size` we need the additional flipping logic because it is dependent on how many of the dimensions were actually specified.

Reply:
Thanks @AaronBallman ! As per your suggestion I have added a commit with an implementation that splits the attribute into two; `ReqdWorkGroupSizeAttr` which should be handled like in upstream and `SYCLReqdWorkGroupSizeAttr` which should have the behavior we need for SYCL. They share the same spellings but the latter is picked if we are in SYCL mode. I will do more cleanup soon, but this should at least give a taste of what it will look like, with a handful of changes actually being reverts from upstream LLVM.

Reply:
I'm largely responsible for much of the complexity in `sycl::reqd_work_group_size`, so I apologize that this has become such a mess.  I assumed that the `sycl::` namespace would be enough to completely isolate the SYCL attribute from any existing OpenCL work.

The good news is that there is a path to deprecation here: [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc) provides equivalent functionality without exposing a user-facing attribute.

Reply:
As discussed offline we will go ahead with the merge of the new variant of the changes. @AaronBallman - Please feel free to do post-commit review of this and we can address it accordingly.